File: bubble_sort/bubble_sort.py
# implementation of bubble sort in Python


def bubble_sort(a_list):
    """Performs bubble sort on a given list.

    a_list -- list of integers

    >>> my_list = [100,6,34,78,2,100,10000,5,789]
    >>> bubble_sort(my_list)
    [2, 5, 6, 34, 78, 100, 100, 789, 10000]
    """

    for pass_num in range(len(a_list) - 1, 0, -1):
        for i in range(pass_num):
            if a_list[i] > a_list[i + 1]:
                a_list[i], a_list[i + 1] = a_list[i + 1], a_list[i]
    return a_list

File: bubble_sort/bubbleSort.js
// iterative implementation of bubble sort in JS on an array of integers

const arr = [100, 6, 34, 78, 2, 100, 10000, 5, 789];

const bubbleSort = (arr) => {
  arr.forEach(() => {
    for (let i = 0; i < arr.length; i++) {
      if (arr[i] > arr[i + 1]) {
        [arr[i], arr[i+1]] = [arr[i+1], arr[i]];
      }
    }
  });
  return arr;
}

console.log(bubbleSort(arr));


// recursive implementation of bubble sort in JS on an array of integers
const sort = (arr, n) => {
  if (n === 1) {
    return arr;
  }
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] > arr[i + 1]) {
      [arr[i], arr[i + 1]] = [arr[i + 1], arr[i]];
    }
  }
  return sort(arr, n - 1);
};

const recursiveBubbleSort = (vector) => {
  const n = vector.length - 1;
  return sort(vector, n);
}

console.log(recursiveBubbleSort(arr, arr.length - 1));
